File: backend/app/url_guard.py
"""Outbound URL validation for backend fetch and proxy endpoints."""
import ipaddress
import socket
from urllib.parse import urljoin, urlparse

import httpx


def _public_url(value):
    parsed = urlparse(value)
    if parsed.scheme not in ('http', 'https') or not parsed.hostname or parsed.username or parsed.password:
        raise ValueError('only public http(s) URLs are allowed')
    try:
        addresses = socket.getaddrinfo(parsed.hostname, parsed.port or (443 if parsed.scheme == 'https' else 80), type=socket.SOCK_STREAM)
    except OSError as err:
        raise ValueError('URL host could not be resolved') from err
    for address in {result[4][0] for result in addresses}:
        ip = ipaddress.ip_address(address)
        if not ip.is_global:
            raise ValueError('private or non-public URL destinations are not allowed')
    return value


async def safe_get(client, url, *, stream=False, max_redirects=5):
    """GET a public URL while validating every redirect destination."""
    current = _public_url(url)
    for _ in range(max_redirects + 1):
        request = client.build_request('GET', current)
        response = await client.send(request, stream=stream, follow_redirects=False)
        if response.status_code not in (301, 302, 303, 307, 308):
            return response
        location = response.headers.get('location')
        if not location:
            return response
        await response.aclose()
        current = _public_url(urljoin(current, location))
    raise ValueError('too many redirects')

File: rust/server/src/url_guard.rs
//! SSRF protection — validate outbound URLs are public.

use std::net::{IpAddr, ToSocketAddrs};
use url::Url;

pub fn public_url(value: &str) -> Result<String, String> {
    let parsed = Url::parse(value).map_err(|_| "invalid URL".to_string())?;
    let scheme = parsed.scheme();
    if scheme != "http" && scheme != "https" {
        return Err("only public http(s) URLs are allowed".to_string());
    }
    let host = parsed.host_str().ok_or("URL has no hostname".to_string())?;
    if parsed.username() != "" || parsed.password().is_some() {
        return Err("only public http(s) URLs are allowed".to_string());
    }
    let port = parsed.port_or_known_default().unwrap_or(if scheme == "https" { 443 } else { 80 });
    let addrs: Vec<IpAddr> = format!("{}:{}", host, port)
        .to_socket_addrs()
        .map_err(|_| "URL host could not be resolved".to_string())?
        .map(|sa| sa.ip())
        .collect();
    for addr in &addrs {
        if !is_global(addr) {
            return Err("private or non-public URL destinations are not allowed".to_string());
        }
    }
    Ok(value.to_string())
}

fn is_global(ip: &IpAddr) -> bool {
    match ip {
        IpAddr::V4(v4) => {
            !v4.is_private() && !v4.is_loopback() && !v4.is_link_local()
                && !v4.is_broadcast() && !v4.is_unspecified()
                && !v4.is_documentation()
        }
        IpAddr::V6(v6) => {
            !v6.is_loopback() && !v6.is_unspecified()
            // Rough check: skip unique-local (fc00::/7)
            && !(v6.segments()[0] & 0xfe00 == 0xfc00)
        }
    }
}

pub async fn safe_get(client: &reqwest::Client, url: &str, max_redirects: usize) -> Result<reqwest::Response, String> {
    let mut current = public_url(url)?;
    for _ in 0..=max_redirects {
        let resp = client.get(&current)
            .send().await
            .map_err(|e| format!("request failed: {}", e))?;
        let status = resp.status().as_u16();
        if ![301, 302, 303, 307, 308].contains(&status) {
            return Ok(resp);
        }
        let location = resp.headers().get("location")
            .and_then(|v| v.to_str().ok())
            .ok_or("redirect with no location")?;
        let base = Url::parse(&current).map_err(|_| "invalid URL".to_string())?;
        current = base.join(location).map_err(|_| "invalid redirect URL".to_string())?.to_string();
        public_url(&current)?;
    }
    Err("too many redirects".to_string())
}
